crypto/cms/cms_lib.c

```diff
@@ -431,6 +433,22 @@ BIO *ossl_cms_DigestAlgorithm_init_bio(X509_ALGOR *digestAlgorithm,
         ERR_raise(ERR_LIB_CMS, CMS_R_MD_BIO_INIT_ERROR);
         goto err;
     }
+    if (EVP_MD_is_a(digest, SN_shake128))
+        xof_len = 32;
+    else if (EVP_MD_is_a(digest, SN_shake256))
+        xof_len = 64;
+    if (xof_len > 0) {
+        EVP_MD_CTX *mdctx;
+        OSSL_PARAM params[2];
+
+        if (BIO_get_md_ctx(mdbio, &mdctx) <= 0 || mdctx == NULL)
+            goto err;
+        params[0] = OSSL_PARAM_construct_size_t(OSSL_DIGEST_PARAM_XOFLEN,
+                                                &xof_len);
+        params[1] = OSSL_PARAM_construct_end();
+        if (!EVP_MD_CTX_set_params(mdctx, params))
+            goto err;
+    }
```

Comment:
I think this would be worthwhile guarding with a check that the digest is an XOF one.  The `EVP_MD_is_a` are fairly expensive.

```suggestion
    if (EVP_MD_xof(digest)) {
        if (EVP_MD_is_a(digest, SN_shake128))
            xof_len = 32;
        else if (EVP_MD_is_a(digest, SN_shake256))
            xof_len = 64;
        if (xof_len > 0) {
            EVP_MD_CTX *mdctx;
            OSSL_PARAM params[2];

            if (BIO_get_md_ctx(mdbio, &mdctx) <= 0 || mdctx == NULL)
                goto err;
            params[0] = OSSL_PARAM_construct_size_t(OSSL_DIGEST_PARAM_XOFLEN,
                                                    &xof_len);
            params[1] = OSSL_PARAM_construct_end();
            if (!EVP_MD_CTX_set_params(mdctx, params))
                goto err;
        }
    }
```